### PR TITLE
checkpoint: support etcd checkpoint

### PR DESCRIPTION
**Pull request by xiang90**
/cc @aaronlevy @derekparker 

Any idea about how to test this?

### CODE REVIEW

**Comment by aaronlevy** [comment]
One option for testing:

1. Build a checkpoint image from this PR (there is a private jenkins job you can use to do this)
2. Launch a cluster using https://github.com/kubernetes-incubator/bootkube/tree/multi-node-self-host-etcd -- but change the image repo:tag to reference the image built in step 1 (https://github.com/kubernetes-incubator/bootkube/blob/master/pkg/asset/internal/templates.go#L184)
3. restart the node running etcd

In terms of general functionality -- I'd like to move to using pod annotations rather than searching for fixed pod names. Also, just to note, we also will need to implement checkpointed pod GC for this to work well with etcd.

https://github.com/kubernetes-incubator/bootkube/issues/205
https://github.com/kubernetes-incubator/bootkube/issues/183

**Comment by xiang90** [comment]
> I'd like to move to using pod annotations rather than searching for fixed pod names. 

Do you want me to change this in this pr or a separate one? I feel adding etcd checkpointer and using annotation is orthogonal. But I can make the matching to use annotation first.

File: cmd/checkpoint/main.go [hunk 1]
```diff
@@ -29,6 +30,9 @@ const (
 
 	tempAPIServer = "temp-apiserver"
 	kubeAPIServer = "kube-apiserver"
+
+	tempEtcd   = "temp-self-hosted-etcd"
+	etcdPrefix = "self-hosted-etcd"
```
Comment (xiang90) [inline cmd/checkpoint/main.go]: Sure.

File: cmd/checkpoint/main.go [hunk 2]
```diff
@@ -46,12 +50,32 @@ var tempAPIServerManifest = v1.Pod{
...
+}
+
 var tempPodSpecMap = map[string]v1.Pod{
 	tempAPIServer: tempAPIServerManifest,
+	tempEtcd:      tempEtcdManifest,
 }
 
 func main() {
 	glog.Info("begin pods checkpointing...")
+	etcd := flag.Bool("experimental-self-hosted-etcd", false, "Checkpoint self hosted etcd.")
```
Comment (xiang90) [inline cmd/checkpoint/main.go]: If that is the case, I think it makes sense to finish the annotation thing first? Then the checkpointer knows what to pick up to checkpoint.

Comment (aaronlevy) [inline cmd/checkpoint/main.go]: That would be great if you wanted to make that change.

Comment (aaronlevy) [inline cmd/checkpoint/main.go]: Or we could get by with it as-is, just without the flag (checkpoint etcd pod by name if it exists)

Comment (xiang90) [inline cmd/checkpoint/main.go]: Let's me try it first tomorrow.

**Comment by aaronlevy** [comment]
Closing, as generic pod checkpointing was implemented in https://github.com/kubernetes-incubator/bootkube/pull/238

**Pull request closed without merging**